### ISSUE TITLE
Automatic authentication from json file - Error in pubkey_type(derive_pubkey(key)) : Unsupported key type: 0

### ISSUE THREAD

**Issue by tomsing1** [cloudyr/googleCloudStorageR#53]
I am trying to authenticate from a json file, downloaded from the Google Project following [your instructions ](https://github.com/cloudyr/googleCloudStorageR#auto-authentication). 

I downloaded the keys for a service account of the project that owns the cloud bucket. The location of this json file is available in the GCS_AUTH_FILE system environment variable. 

For example, I load the R package with the following code:
```
Sys.setenv("GCS_AUTH_FILE" ="my_key.json")
library(googleCloudStorageR)
```
but I get the following error:

```
Error : .onAttach failed in attachNamespace() for 'googleCloudStorageR', details:
  call: pubkey_type(derive_pubkey(key))
  error: Unsupported key type: 0
Error: package or namespace load failed for ‘googleCloudStorageR’
```

I think the json file is found, but either it doesn't have the right content or something else isn't quite right. Any hints on where I might be going wrong? (Perhaps you have an example (without the private information, obviously) of what the json file should contain?)

In general, my goal is to create an R markdown script that can be run without manual authentication, eg by pressing the "Knit HTML" button in RStudio. As always, any advice is greatly appreciated.

SessionInfo:
```
R version 3.2.2 (2015-08-14)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] switchr_0.9.19

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.6                    googleCloudStorageR_0.1.0.9000 crayon_1.3.2                   digest_0.6.10                 
 [5] withr_1.0.2                    mime_0.5                       R6_2.1.3                       jsonlite_1.0                  
 [9] xtable_1.8-2                   magrittr_1.5                   httr_1.2.1                     testthat_1.0.2                
[13] googleAuthR_0.4.0              devtools_1.12.0                tools_3.2.2                    shiny_0.13.2                  
[17] httpuv_1.3.3                   memoise_1.0.0                  htmltools_0.3.5                openssl_0.9.4                 
[21] knitr_1.14                
```

**Comment by MarkEdmondson1234** [comment]
It is confusing which JSON to download in the Google project, could you perhaps post an edited output of the JSON file content? (keep private keys secret) 

It needs to be the one that has the private_key field, which is from "Service Account Keys" in the Google API console, not "service account client".

**Comment by tomsing1** [comment]
Sure, here is an (edited) version of my json file:

```
library(jsonlite)
fromJSON(GCS_AUTH_FILE)
```

```
$type
[1] "service_account"

$project_id
[1] "myproject-12"

$private_key_id
[1] "EDITED_OUT"

$private_key
[1] "-----BEGIN PRIVATE KEY-----\nEDITED_OUT\n-----END PRIVATE KEY-----\n"

$client_email
[1] "thomas@myproject-12.iam.gserviceaccount.com"

$client_id
[1] "EDITED_OUT"

$auth_uri
[1] "https://accounts.google.com/o/oauth2/auth"

$token_uri
[1] "https://accounts.google.com/o/oauth2/token"

$auth_provider_x509_cert_url
[1] "https://www.googleapis.com/oauth2/v1/certs"

$client_x509_cert_url
[1] "https://www.googleapis.com/robot/v1/metadata/x509/thomas%40myproject-12.iam.gserviceaccount.com"

```

**Comment by MarkEdmondson1234** [comment]
Hmm, looks ok.

Can you run googleAuthR::gar_auth_service("your.json") and see if that works?

Also check you have the right API turned on for your project. 

And are you setting your own project client ID and client secret via environment vars too? I may have not stressed that in the docs.

**Comment by tomsing1** [comment]
Thanks for the quick response. Here is what I get:

```
googleAuthR::gar_auth_service("my_key.json")
Error in pubkey_type(derive_pubkey(key)) : Unsupported key type: 0
```

When I manually authenticate using `gcs_auth()` I can successfully upload files to cloud storage, so the API seems to be active.

I have not set project client ID and client secret via environment variables. Would that be necessary, given that they key / secret are already available in the json file?

**Comment by MarkEdmondson1234** [comment]
Yes you will need to set the options I think to be the same as in the JSON file, although now you say it it does make sense to read it from the file instead, lets at leasst see if that is the issue:

Run the below before loading the library or authenticating.

```r
options("googleAuthR.client_id" = "YOUR_CLIENT_ID")
options("googleAuthR.client_secret" = "YOUR_CLIENT_SECRET")
```

**Comment by tomsing1** [comment]
Sorry, still not luck. I executed the following code, with `path_to_json` pointing to the json file I obtained from the Cloud project for the service account.

```
library(jsonlite)
gcs.key <- fromJSON(path_to_json)
Sys.setenv("GCS_AUTH_FILE" = path_to_json)
options("googleAuthR.client_id" = gcs.key$client_id)
options("googleAuthR.client_secret" = gcs.key$private_key)
library(googleCloudStorageR) 
```

The following error is returned:

```
Setting scopes to https://www.googleapis.com/auth/devstorage.full_control
If you need additional scopes set do so via options(googleAuthR.scopes.selected = c('scope1', 'scope2')) before loading library and include one required scope.
Error : .onAttach failed in attachNamespace() for 'googleCloudStorageR', details:
  call: pubkey_type(derive_pubkey(key))
  error: Unsupported key type: 0
In addition: Warning message:
replacing previous import by ‘googleAuthR::gar_api_generator’ when loading ‘googleCloudStorageR’ 
Error: package or namespace load failed for ‘googleCloudStorageR’
```

Anything else I could try?

```
> sessionInfo()
R version 3.2.2 (2015-08-14)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] jsonlite_1.0   switchr_0.9.19

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.6                    magrittr_1.5                   googleCloudStorageR_0.1.0.9000 devtools_1.12.0               
 [5] xtable_1.8-2                   R6_2.1.3                       httr_1.2.1                     tools_3.2.2                   
 [9] googleAuthR_0.4.0              withr_1.0.2                    htmltools_0.3.5                openssl_0.9.4                 
[13] digest_0.6.10                  crayon_1.3.2                   shiny_0.13.2                   testthat_1.0.2                
[17] memoise_1.0.0                  mime_0.5                       httpuv_1.3.3     
```

**Comment by MarkEdmondson1234** [comment]
I checked and the `private_key` in the JSON file is not the same as the `client secret` in the Google Console - you need to read those out from the Google Console (from API Manager > Credentials > OAuth 2.0 Client IDs > click on your project name 

Also you can check using `googleAuthR::gar_auth_service()` as that is the function that is being called.

The error message you are seeing is when the private key can't be decrypted, which does suggest the secret isn't being read in the methods tried above.

This should be the best script to check it:

```r
options(googleAuthR.client_id = "CLIENT ID FROM GOOGLE CONSOLE")
options(googleAuthR.client_secret = "CLIENT SECRET FROM GOOGLE CONSOLE")
library(googleAuthR)
gar_auth_service("location_of_json.json")
```

If the above succeeds then the auto-auth should work.

Thanks for this, at the very least it helps with where the documentation should be updated, I'll also look at more informative error messages should this be the issue.

**Comment by tomsing1** [comment]
Thanks a ton for your detailed response. Unfortunately, I still haven't been able to authenticate with the service account. To figure out where I am doing something wrong, here are all of the steps I perform:

1. Browsed to API-Manager > Credentials
2. Clicked "Create credentials" > OAuth client ID > Other and noted the
    client ID and client secret.
3. Set the GCS_CLIENT_ID and GCS_CLIENT_SECRET environmental variables accordingly:

```Sys.setenv(
  "GCS_CLIENT_ID" = "XXXX.apps.googleusercontent.com",
  "GCS_CLIENT_SECRET" = "XXXX")
```
4. Browsed to API-Manager > Credentials
5. Clicked "Create credentials" > "Service account key" 
6. Selected "App engine default service account" and downloaded a json file
   with the keys. (I also tried the Compute engine default service account.)
8. Loaded the googleAuthR package with a library call.
9. `gar_auth_service("json_file_from_two_steps_before.json")`

```
Error in pubkey_type(derive_pubkey(key)) : Unsupported key type: 0
```

Thanks for taking another look. By the way, I can authenticate interactively via `gcs_auth()`, store the `.httr-oauth` file in a location of my choice and then authenticate automatically again by setting   `Sys.setenv("GCS_AUTH_FILE" = "/fullpath/to/.httr.oauth")` before loading the `googleCloudStorageR` package. That's great already and might be my solution going forward.

**Comment by MarkEdmondson1234** [comment]
Ok cool, glad you have a work-around.  I'd like to get to the bottom of it as it is cool and I want to use it for other packages.  I use a JSON file myself and it works fine, so I'll leave the issue open  and hope someone else gets the issue and we can help debug it :)

**Comment by MarkEdmondson1234** [comment]
I see its on linux - could I check if you have `libssl-dev` installed?  I'm taking this from a dependency package [openssl readme](https://github.com/jeroenooms/openssl) that handles this decryption.

`sudo apt-get install -y libssl-dev`

**Comment by tomsing1** [comment]
Thanks again for your help. I have now successfully authenticated with a json file for a service account on a `Mac OS X` laptop and a `Ubuntu 14.04 Linux VM` in the Google cloud.
It worked without a hitch in both cases. 

Just for completeness, here's what I did step by step:

1. Set up a service account and got the JSON file as documented here:
   https://cloud.google.com/storage/docs/authentication#generating-a-private-key
2. Within R (replacing PATH_TO_MY_JSON_FILE and MY_BUCKET with the right paths):
 
```
  install.packages("devtools")
  library(devtools)
  install_github("cloudyr/googleCloudStorageR")
  Sys.setenv("GCS_AUTH_FILE" = "PATH_TO_MY_JSON_FILE.json")
  library(googleCloudStorageR)
  gcs_list_buckets("MY_BUCKET")
```

So it seems that my local linux workstation might be to blame for the troubles I reported. Your suspicion that I might not have the right openSSL library is likely the root cause (but I can't change and test that right now). Case closed!

By the way, I found [the above documentation](https://cloud.google.com/storage/docs/authentication#generating-a-private-key) really helpful to set up a service account and get the right json file for the service account.
Perhaps you want to link to it in your documentation?

Again, many thanks for helping me trouble shoot and for making these great packages available to the community!
